### PR TITLE
[SPARK-43790][PYTHON][CONNECT][ML] Add `copyFromLocalToFs` API

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

Added `copyFromLocalToFs` API.
The API copies a provided local file to provided destination path on hadoop FS.

The implementation of the API reuses `AddArtifact` proto message, and it fill the Artifact path as `forward_to_fs/{destination_path}` , and in spark connect server backend, when handling artifact requests, if receiving an artifact request with path starting with `forward_to_fs`, it saves the artifact data to the provided `destination_path` on HadoopFS.

### Why are the changes needed?
In new distributed spark ML module (designed to support spark connect and support local inference)

We need to save ML model to hadoop file system as its raw file format and saved as provided file name, the reason is:

We often submit a spark application to spark cluster for running the training model job, we need to save trained model to hadoop file system before the spark application completes.
But we want to support local model inference, that means if we save the model by current spark DataFrame writer (e.g. parquet format), when loading model we have to rely on the spark service. But we hope we can load model without spark service. So we want the model being saved as the original binary format that our ML code can handle.

### Does this PR introduce _any_ user-facing change?
Yes

Added `copyFromLocalToFs` API:

```
    def copyFromLocalToFs(self, local_path: str, dest_path: str) -> None:
        """
        copy file from local to FS.

        Parameters
        ----------
        local_path: str
            Path to a local file (Directory is not supported yet).
            The path can be either absolute path or relative path.

        dest_path: str
            The FS path to the destination the file will be copied to.

        .. versionadded:: 3.5.0
        """
```

### How was this patch tested?

Unit tests.

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala [hunk 1]
```diff
@@ -89,6 +92,28 @@ class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddAr
...
+      val fs = destFSPath.getFileSystem(hadoopConf)
+      fs.copyFromLocalFile(
+        false,
+        true,
+        new FSPath(localPath.toString),
+        destFSPath
+      )
+    } catch {
+      case e: java.io.IOException =>
+        // TODO: report failure in response message
```
Comment (WeichenXu123) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:112]: Discussion: Shall we add a field in ArtifactResponse message for holding failure information here ?

File: python/pyspark/sql/connect/client/core.py [hunk 2]
```diff
@@ -1240,6 +1240,9 @@ def _handle_rpc_error(self, rpc_error: grpc.RpcError) -> NoReturn:
     def add_artifacts(self, *path: str, pyfile: bool, archive: bool) -> None:
         self._artifact_manager.add_artifacts(*path, pyfile=pyfile, archive=archive)
 
+    def copy_from_local_to_FS(self, local_path: str, dest_path: str) -> None:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/core.py:1243]: `copy_from_local_to_FS` -> `copy_from_local_to_fs`

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala (hunk 1 above)
Comment (HyukjinKwon) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:112]: I believe @allisonwang-db is working on that.

File: python/pyspark/sql/tests/connect/client/test_artifact.py [hunk 3]
```diff
@@ -271,6 +271,21 @@ def func(x):
             self.spark.addArtifacts(f"{archive_path}.zip#my_files", archive=True)
             self.assertEqual(self.spark.range(1).select(func("id")).first()[0], "hello world!")
 
+    def test_copy_from_local_to_FS(self):
```
Comment (vicennial) [inline python/pyspark/sql/tests/connect/client/test_artifact.py:274]: Please add a server/scala-side test as well in `ArtifactManagerSuite`

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala (hunk 1 above)
Comment (allisonwang-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:112]: I think we should catch all gRPC errors using this handleError: https://github.com/apache/spark/blob/7d87fecda7077714601cdf7328a11d2719a99bb1/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala#L119-L123

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala [hunk 4]
```diff
@@ -89,6 +92,28 @@ class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddAr
     artifactManager.addArtifact(holder, artifact.path, artifact.stagedPath, artifact.fragment)
   }
 
+  protected def uploadStagedArtifactToFS(artifact: StagedArtifact): Unit = {
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:95]: No problem. :)

File: python/pyspark/sql/tests/connect/client/test_artifact.py (hunk 3 above)
Comment (WeichenXu123) [reply python/pyspark/sql/tests/connect/client/test_artifact.py:274]: Sure.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala (hunk 1 above)
Comment (WeichenXu123) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:112]: @allisonwang-db 

`handleError` is defined in `SparkConnectService` and it is private, but my code is in `SparkConnectAddArtifactsHandler` class, how can I call `handleError` from `SparkConnectAddArtifactsHandler` code ? Could you make `handleError` to be a utility function that is accessible form other classes ?

Comment (WeichenXu123) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:112]: `SparkConnectAddArtifactsHandler` already defined `onError` handler, see

https://github.com/apache/spark/blob/e8f87413bdd905e5e208ac8484b3b11de1623fc2/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala#L83

We might need to do some refactor to make it use `handleError` , @allisonwang-db would you mind to do it in follow-up PR ?

Comment (allisonwang-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala]: @MaxGekk will be working on a new framework for error handling. @WeichenXu123 You can migrate to use the new framework when it's completed.

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/artifact/ArtifactManagerSuite.scala [hunk 5]
```diff
@@ -145,4 +145,23 @@ class ArtifactManagerSuite extends SharedSparkSession with ResourceHelper {
       assert(artifactManager.getSparkConnectPythonIncludes == Seq("abc.zip"))
     }
   }
+
+  test("Forward artifact file to cloud storage path") {
```
Comment (HyukjinKwon) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/artifact/ArtifactManagerSuite.scala:149]:
```suggestion
  test("SPARK-43790: Forward artifact file to cloud storage path") {
```

File: python/pyspark/sql/connect/session.py [hunk 6]
```diff
@@ -625,6 +625,26 @@ def addArtifacts(self, *path: str, pyfile: bool = False, archive: bool = False)
 
     addArtifact = addArtifacts
 
+    def copyFromLocalToFs(self, local_path: str, dest_path: str) -> None:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:628]: Can be move this to ML package? Should also mark it as a developer API in `Notes` section.

File: python/pyspark/sql/connect/client/artifact.py [hunk 7]
```diff
@@ -187,6 +188,19 @@ def _parse_artifacts(self, path_or_uri: str, pyfile: bool, archive: bool) -> Lis
             return [artifact]
         raise RuntimeError(f"Unsupported scheme: {parsed.scheme}")
 
+    def _parse_forward_to_FS_artifacts(self, local_path: str, dest_path: str) -> List[Artifact]:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/artifact.py:191]:
```suggestion
    def _parse_forward_to_fs_artifacts(self, local_path: str, dest_path: str) -> List[Artifact]:
```

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala [hunk 8]
```diff
@@ -157,10 +159,34 @@ class SparkConnectArtifactManager private[connect] {
       }
     }
   }
+
+  private[connect] def uploadArtifactToFs(
+      sessionHolder: SessionHolder,
+      remoteRelativePath: Path,
```
Comment (HyukjinKwon) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:165]: I believe it's not a relative path?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala [hunk 9]
```diff
@@ -100,7 +100,15 @@ class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddAr
       // We do not store artifacts that fail the CRC. The failure is reported in the artifact
       // summary and it is up to the client to decide whether to retry sending the artifact.
       if (artifact.getCrcStatus.contains(true)) {
-        addStagedArtifactToArtifactManager(artifact)
+        if (artifact.path.startsWith(
```
Comment (HyukjinKwon) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:103]: Hm, shouldn't we do this at `addStagedArtifactToArtifactManager`?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala [hunk 10]
```diff
@@ -157,10 +159,34 @@ class SparkConnectArtifactManager private[connect] {
...
+      serverLocalStagingPath: Path): Unit = {
+    val hadoopConf = sessionHolder.session.sparkContext.hadoopConfiguration
+    assert(remoteRelativePath.startsWith(
+      SparkConnectArtifactManager.forwardToFSPrefix + File.separator)
```
Comment (HyukjinKwon) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:169]: I am fine with this but one thing is that now here we pass an absolute path with the prefix. Except here, all other places pass a relative path around. Is that fine, @hvanhovell and @vicennial?

File: python/pyspark/sql/connect/client/artifact.py (hunk 7 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/client/artifact.py:191]: BTW, this one too. Shouldn't we do it together with `_parse_artifacts`?

Comment (WeichenXu123) [reply python/pyspark/sql/connect/client/artifact.py:191]: I don't think so. The arguments are different. For `_parse_forward_to_FS_artifacts`, we have `local_path` and `dest_path` and the 2 paths are different.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala (hunk 10 above)
Comment (WeichenXu123) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:169]: In the design suggested by @vicennial , We defined the artifact path to be `forward_to_fs/{dest_fs_path}`, and `dest_fs_path` is absolute path to cloud storage. I think it is fine and it works in all cases.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala (hunk 9 above)
Comment (WeichenXu123) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:103]: The goal for artifact with `forward_to_fs/{dest_fs_path}` path is just to upload file to cloud storage. Do we still need add the artifact file to artifact manager ? 
If you want , I can add it, it is simple.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala [hunk 11]
```diff
@@ -157,10 +159,34 @@ class SparkConnectArtifactManager private[connect] {
...
+      Paths.get("/").resolve(
+        remoteRelativePath.subpath(1, remoteRelativePath.getNameCount)).toString
+    )
+    val localPath = serverLocalStagingPath
+    val fs = destFSPath.getFileSystem(hadoopConf)
+    fs.copyFromLocalFile(
+      false,
+      true,
+      new FSPath(localPath.toString),
+      destFSPath
```
Comment (WeichenXu123) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:181]: Save file to cloud storage file system.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala (hunk 10 above)
Comment (HyukjinKwon) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:169]: 🆗

**Comment by HyukjinKwon** [comment]
Looks fine to me but I would leave it to @hvanhovell and @vicennial to check and merge.

File: python/pyspark/sql/connect/session.py (hunk 6 above)
Comment (WeichenXu123) [reply python/pyspark/sql/connect/session.py:628]: I feel like this API is useful in many places, not only for ML, what about keeps putting it  here ? We can still mark it as developer API

File: python/pyspark/sql/connect/client/artifact.py [hunk 12]
```diff
@@ -187,6 +188,19 @@ def _parse_artifacts(self, path_or_uri: str, pyfile: bool, archive: bool) -> Lis
             return [artifact]
         raise RuntimeError(f"Unsupported scheme: {parsed.scheme}")
 
+    def _parse_forward_to_fs_artifacts(self, local_path: str, dest_path: str) -> List[Artifact]:
+        local_path = Path(local_path).absolute()
+        # TODO: Support directory path.
+        assert local_path.is_file(), "local path must be a file path."
+        storage = LocalFile(str(local_path))
+
+        assert Path(dest_path).is_absolute(), "destination FS path must be an absolut path."
```
Comment (vicennial) [inline python/pyspark/sql/connect/client/artifact.py:197]: Nit: Typo `absolut`

**Review by vicennial — APPROVED** [review]
LGTM once tests pass

**Review by vicennial — APPROVED** [review]
LGTM once tests pass

**Review by vicennial — APPROVED** [review]
LGTM once tests pass

**Review by vicennial — APPROVED** [review]
LGTM once tests pass

**Review by vicennial — APPROVED** [review]
LGTM once tests pass

**Review by vicennial — APPROVED** [review]
LGTM once tests pass

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala [hunk 13]
```diff
@@ -157,10 +159,46 @@ class SparkConnectArtifactManager private[connect] {
...
+    val destFSPath = new FSPath(
+      Paths
+        .get("/")
+        .resolve(remoteRelativePath.subpath(1, remoteRelativePath.getNameCount))
+        .toString)
+    val localPath = serverLocalStagingPath
+    val fs = destFSPath.getFileSystem(hadoopConf)
+    if (fs.isInstanceOf[LocalFileSystem]) {
+      val allowDestLocalConf = sessionHolder.session.sparkContext.conf
+        .get("spark.fs.copyFromLocalToFs.allowDestLocal", "false")
```
Comment (WeichenXu123) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:180]: Used in testing.

Comment (HyukjinKwon) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/artifact/SparkConnectArtifactManager.scala:180]: I think we should better add this conf to `org.apache.spark.sql.connect.config.Connect`, and explicitly document that it's for testing purpose.

File: python/pyspark/sql/connect/session.py (hunk 6 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/connect/session.py:641]: Can we add `.. versionadded 3.5.0` directive?

File: python/pyspark/sql/connect/session.py [hunk 14]
```diff
@@ -637,6 +638,34 @@ def addArtifacts(self, *path: str, pyfile: bool = False, archive: bool = False)
...
+        """
+        Copy file from local to cloud storage file system.
+        If the file already exits in destination path, old file is overwritten.
+
+        Parameters
+        ----------
+        local_path: str
+            Path to a local file (Directory is not supported yet).
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:649]:
```suggestion
            Path to a local file. Directories are not supported.
```

File: python/pyspark/sql/tests/connect/client/test_artifact.py [hunk 15]
```diff
@@ -271,6 +277,21 @@ def func(x):
...
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/client/test_artifact.py:280]:
```suggestion
    def test_copy_from_local_to_fs(self):
```

File: python/pyspark/sql/connect/session.py [hunk 16]
```diff
@@ -637,6 +638,34 @@ def addArtifacts(self, *path: str, pyfile: bool = False, archive: bool = False)
...
+
+        Notes
+        -----
+        This API is a developer API.
+        """
+        if urllib.parse.urlparse(dest_path).scheme:
+            raise ValueError(
+                "`spark_session.copyFromLocalToFs` API only allows `dest_path` to be a path "
+                "without scheme, and spark driver uses the default scheme to "
+                "determine the destination file system."
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:665]: Why can't we support the fully qualified URI with the scheme? We just add a prefix and truncate it later in the server.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/session.py [hunk 17]
```diff
@@ -642,6 +643,34 @@ def addArtifacts(
...
+            The path can be either an absolute path or a relative path.
+
+        dest_path: str
+            The cloud storage path to the destination the file will
+            be copied to.
+            The path must be an an absolute path.
+
+        Notes
+        -----
+        This API is a developer API.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/session.py:664]: Seems like we missed versionadded directive, @WeichenXu123

**Pull request closed without merging**